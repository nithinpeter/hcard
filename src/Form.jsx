import React from 'react';
import PropTypes from 'prop-types';
import UploadAvatar from './UploadAvatar';
import { commonPropTypes } from './common';

import './Form.scss';

const Form = props => {
  return (
    <div className="form">
      <h1 className="form__title">hCard Builder</h1>
      <form action="">
        <fieldset className="form__fieldset">
          <legend className="form__fieldset__legend">Personal details</legend>
          <div className="form__fieldset__fields">
            <div className="form__fieldset__fields__row">
              <Field
                id="givenName"
                value={props.givenName}
                label="Given name"
                onChange={props.onChange}
              />
              <Field
                id="surname"
                value={props.surname}
                label="Surname"
                onChange={props.onChange}
              />
            </div>
            <div className="form__fieldset__fields__row">
              <Field
                id="email"
                value={props.email}
                label="Email"
                onChange={props.onChange}
                type="email"
              />
              <Field
                id="phone"
                value={props.phone}
                label="Phone"
                onChange={props.onChange}
                type="tel"
              />
            </div>
          </div>
        </fieldset>
        <fieldset className="form__fieldset">
          <legend className="form__fieldset__legend">Address</legend>
          <div className="form__fieldset__fields">
            <div className="form__fieldset__fields__row">
              <Field
                id="houseName"
                value={props.houseName}
                label="House name or #"
                onChange={props.onChange}
              />
              <Field
                id="street"
                value={props.street}
                label="Street"
                onChange={props.onChange}
              />
            </div>
            <div className="form__fieldset__fields__row">
              <Field
                id="suburb"
                value={props.suburb}
                label="Suburb"
                onChange={props.onChange}
              />
              <Field
                id="state"
                value={props.state}
                label="State"
                onChange={props.onChange}
              />
            </div>
            <div className="form__fieldset__fields__row">
              <Field
                id="postCode"
                value={props.postCode}
                label="Postcode"
                onChange={props.onChange}
              />
              <Field
                id="country"
                value={props.country}
                label="Country"
                onChange={props.onChange}
              />
            </div>
          </div>
        </fieldset>
        <div className="form__buttons">
          <div className="form__buttons__btn">
            <UploadAvatar onChange={props.onChange} />
          </div>
          <div className="form__buttons__btn">
            <button type="submit" className="form__buttons__btn--submit">
              Submit
            </button>
          </div>
        </div>
      </form>
    </div>
  );
};

Form.propTyes = {
  ...commonPropTypes,
  onChange: PropTypes.func.isRequired,
}

const Field = ({ id, value, label, type, onChange }) => (
  <div className="field">
    <label htmlFor={id} className="field__label">
      {label}
    </label>
    <input
      className="field__input"
      type={type}
      value={value}
      id={id}
      onChange={e => onChange(id, e.target.value)}
    />
  </div>
);

Field.defaultProps = {
  type: 'text',
};

Field.propTypes = {
  id: PropTypes.string.isRequired,
  value: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

export default Form;
