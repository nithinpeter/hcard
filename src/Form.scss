@import './common';

.form {
  flex: 1 0 50%;
  padding: 40px;

  &__title {
    margin-top: 0;
    color: $primary;
  }

  &__fieldset {
    border: none;
    padding: 0;
    margin: 0 0 20px 0;

    &__legend {
      width: 100%;
      font-size: 12px;
      text-transform: uppercase;
      color: $gray-1;
      padding: 2px 0;
      margin: 0 0 20px 0;
      border-bottom: 1px solid $gray-1;
    }

    &__fields {
      display: flex;
      flex-direction: column;

      &__row {
        display: flex;

        @media (max-width: 600px) {
          flex-direction: column;
        }
      }
    }
  }

  &__buttons {
    display: flex;

    @media (max-width: 600px) {
      flex-direction: column;
    }

    &__btn {
      flex-basis: 50%;

      &--submit {
        @include button($primary);
      }

      &:nth-child(1) {
        padding-right: 10px;
      }
      &:nth-child(2) {
        padding-left: 10px;
      }

      @media (max-width: 600px) {
        &:nth-child(1) {
          padding-right: 0;
          margin-bottom: 10px;
        }
        &:nth-child(2) {
          padding-left: 0;
        }
      }
    }
  }
}

.field {
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  padding: 10px 0;

  &:nth-child(1) {
    padding-right: 10px;
  }
  &:nth-child(2) {
    padding-left: 10px;
  }

  @media (max-width: 600px) {
    flex-direction: column;

    &:nth-child(1) {
      padding-right: 0;
    }
    &:nth-child(2) {
      padding-left: 0;
    }
  }

  &__label {
    text-transform: uppercase;
    font-size: 14px;
    margin-bottom: 5px;
    color: $text-color;
  }

  &__input {
    border: 1px solid $gray-1;
    border-radius: 2px;
    padding: 5px 10px;
    font-size: 18px;
  }
}
