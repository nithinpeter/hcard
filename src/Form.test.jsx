import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { mount } from 'enzyme';
import Form from './Form';

const props = {
  givenName: 'Sam',
  surname: 'Fairfax',
  email: 'Sam@fairfax.com',
  phone: '012223344',
  houseName: '2',
  street: 'King st',
  suburb: 'Sydney',
  state: 'NSW',
  postCode: '2000',
  country: 'Australia',
};

it('should render Form', () => {
  expect(
    ReactDOMServer.renderToStaticMarkup(
      <Form {...props} onChange={jest.fn()} />
    )
  ).toMatchSnapshot();
});

it('should trigger onChange with correct arguments', () => {
  const onChange = jest.fn();
  const value = 'value';
  const wrapper = mount(<Form {...props} onChange={onChange} />);

  // Ids of all the input elements.
  // Used to find element and trigger `change` event.
  const ids = [
    'givenName',
    'surname',
    'email',
    'phone',
    'houseName',
    'street',
    'suburb',
    'state',
    'postCode',
    'country',
  ];

  ids.forEach(id => {
    const input = wrapper.find(`#${id}`).at(1);
    input.simulate('change', { target: { value } });

    expect(onChange).toHaveBeenCalledWith(id, value);
  });
});
