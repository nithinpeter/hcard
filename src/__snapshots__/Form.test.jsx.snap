// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Form 1`] = `"<div class=\\"form\\"><h1 class=\\"form__title\\">hCard Builder</h1><form action=\\"\\"><fieldset class=\\"form__fieldset\\"><legend class=\\"form__fieldset__legend\\">Personal details</legend><div class=\\"form__fieldset__fields\\"><div class=\\"form__fieldset__fields__row\\"><div class=\\"field\\"><label for=\\"givenName\\" class=\\"field__label\\">Given name</label><input type=\\"text\\" class=\\"field__input\\" value=\\"Sam\\" id=\\"givenName\\"/></div><div class=\\"field\\"><label for=\\"surname\\" class=\\"field__label\\">Surname</label><input type=\\"text\\" class=\\"field__input\\" value=\\"Fairfax\\" id=\\"surname\\"/></div></div><div class=\\"form__fieldset__fields__row\\"><div class=\\"field\\"><label for=\\"email\\" class=\\"field__label\\">Email</label><input type=\\"email\\" class=\\"field__input\\" value=\\"Sam@fairfax.com\\" id=\\"email\\"/></div><div class=\\"field\\"><label for=\\"phone\\" class=\\"field__label\\">Phone</label><input type=\\"tel\\" class=\\"field__input\\" value=\\"012223344\\" id=\\"phone\\"/></div></div></div></fieldset><fieldset class=\\"form__fieldset\\"><legend class=\\"form__fieldset__legend\\">Address</legend><div class=\\"form__fieldset__fields\\"><div class=\\"form__fieldset__fields__row\\"><div class=\\"field\\"><label for=\\"houseName\\" class=\\"field__label\\">House name or #</label><input type=\\"text\\" class=\\"field__input\\" value=\\"2\\" id=\\"houseName\\"/></div><div class=\\"field\\"><label for=\\"street\\" class=\\"field__label\\">Street</label><input type=\\"text\\" class=\\"field__input\\" value=\\"King st\\" id=\\"street\\"/></div></div><div class=\\"form__fieldset__fields__row\\"><div class=\\"field\\"><label for=\\"suburb\\" class=\\"field__label\\">Suburb</label><input type=\\"text\\" class=\\"field__input\\" value=\\"Sydney\\" id=\\"suburb\\"/></div><div class=\\"field\\"><label for=\\"state\\" class=\\"field__label\\">State</label><input type=\\"text\\" class=\\"field__input\\" value=\\"NSW\\" id=\\"state\\"/></div></div><div class=\\"form__fieldset__fields__row\\"><div class=\\"field\\"><label for=\\"postCode\\" class=\\"field__label\\">Postcode</label><input type=\\"text\\" class=\\"field__input\\" value=\\"2000\\" id=\\"postCode\\"/></div><div class=\\"field\\"><label for=\\"country\\" class=\\"field__label\\">Country</label><input type=\\"text\\" class=\\"field__input\\" value=\\"Australia\\" id=\\"country\\"/></div></div></div></fieldset><div class=\\"form__buttons\\"><div class=\\"form__buttons__btn\\"><label class=\\"upload-avatar undefined\\" role=\\"button\\" tabindex=\\"0\\"><input type=\\"file\\" required=\\"\\"/><span>Upload Avatar</span></label></div><div class=\\"form__buttons__btn\\"><button type=\\"submit\\" class=\\"form__buttons__btn--submit\\">Submit</button></div></div></form></div>"`;
